Made `Delayed::Backend::ActiveRecord::Job` class load lazily

willnet opened the pull request:
Fix #239 In Rails, classes like `ActiveRecord::Base` are designed to be loaded lazily. If a gem eagerly loads classes when required, as this gem did, it can cause issues where configurations written in `config/initializers/*.rb` for ActiveRecord are not applied properly. To avoid this, the `ActiveSupport.on_load(:active_record) do ... end` block is used to delay loading the `ActiveRecord::Base` class and `Delayed::Backend::ActiveRecord::Job` class until after Rails has fully initialized.

In this commit, I wrapped the existing `Delayed::Backend::ActiveRecord::Job` class definition in a block to minimize changes. However, it might be better to separate `Delayed::Backend::ActiveRecord::Configuration` into a dedicated file for easier management.

willnet committed:
Made `Delayed::Backend::ActiveRecord::Job` class load lazily

In Rails, classes like `ActiveRecord::Base` are designed to be loaded lazily. If a gem eagerly loads classes when required, as this gem did, it can cause issues where configurations written in `config/initializers/*.rb` for ActiveRecord are not applied properly. To avoid this, the `ActiveSupport.on_load(:active_record) do ... end` block is used to delay loading the `ActiveRecord::Base` class and `Delayed::Backend::ActiveRecord::Job` class until after Rails has fully initialized.

In this commit, I wrapped the existing `Delayed::Backend::ActiveRecord::Job` class definition in a block to minimize changes. However, it might be better to separate `Delayed::Backend::ActiveRecord::Configuration` into a dedicated file for easier management.

willnet committed:
Split the code into separate files for each class

Rubocop issued warnings, so I resolved it by splitting each class into its own file.

```
lib/delayed/backend/active_record.rb:33:7: C: Metrics/BlockLength: Block has too many lines. [122/25]
      ActiveSupport.on_load(:active_record) do ...
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/delayed/backend/active_record.rb:36:9: W: Lint/ConstantDefinitionInBlock: Do not define constants this way within a block.
        class Job < ::ActiveRecord::Base ...
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

willnet committed:
rubocop -a .

Fix the following offenses using RuboCop’s autocorrect feature.

```
Offenses:

lib/delayed/backend/active_record/job.rb:55:15: C: [Correctable] Layout/MultilineMethodCallIndentation: Align .min_priority with ready_to_run(worker.name, max_run_time) on line 54.
              .min_priority
              ^^^^^^^^^^^^^
lib/delayed/backend/active_record/job.rb:56:15: C: [Correctable] Layout/MultilineMethodCallIndentation: Align .max_priority with ready_to_run(worker.name, max_run_time) on line 54.
              .max_priority
              ^^^^^^^^^^^^^
lib/delayed/backend/active_record/job.rb:57:15: C: [Correctable] Layout/MultilineMethodCallIndentation: Align .for_queues with ready_to_run(worker.name, max_run_time) on line 54.
              .for_queues
              ^^^^^^^^^^^
lib/delayed/backend/active_record/job.rb:58:15: C: [Correctable] Layout/MultilineMethodCallIndentation: Align .by_priority with ready_to_run(worker.name, max_run_time) on line 54.
              .by_priority
              ^^^^^^^^^^^^
```

willnet committed:
Add the magic comment from the original file to each split file

pjurewicz commented:
Good job @willnet

flivni reviewed (commented):
I believe you **also** need to wrap delayed_job_active_record:7, i.e.

```ruby
ActiveSupport.on_load(:active_record) do
  Delayed::Worker.backend = :active_record
end
```

This is because `.backend=` setter also has a side-effect of loading active_record.

flivni commented on lib/delayed/backend/active_record.rb:
```diff
@@ -1,213 +1,9 @@
-        yield(configuration)
-      end
-
-      # A job object that is persisted to the database.
```
This comment never found its way into the refactor.

willnet committed:
Add missing delayed execution

`Delayed::Worker.backend = :active_record` runs `Delayed::Backend::ActiveRecord::Job`, so delay the execution until after `ActiveRecord::Base` is loaded

ref: https://github.com/collectiveidea/delayed_job/blob/eaef62d844ea980433b807a496136bbd0a6ea625/lib/delayed/worker.rb#L69

willnet committed:
Restore  comment lines that was lost during refactoring

willnet commented:
@flivni Good catch! I’ve fixed it.

flivni reviewed (approved):
I carefully reviewed the refactor and it was done cleanly. I verified that the gem now loads rails active record lazily making it compatible with rails startup. The updated code works in our code base.

This PR looks good.

albus522 commented:
This prevents DJ from loading correctly until something in the app touches ActiveRecord which does not work. Most importantly the backend would be left as the default until something outside DJ touches ActiveRecord.
